includes/sanitizers/trait-amp-noscript-fallback.php

```diff
@@ -0,0 +1,101 @@
+<?php
+/**
+ * Trait AMP_Noscript_Fallback.
+ *
+ * @package AMP
+ */
+
+/**
+ * Trait AMP_Noscript_Fallback
+ *
+ * @since 1.1
+ *
+ * Used for sanitizers that place <noscript> tags with the original nodes on error.
+ */
+trait AMP_Noscript_Fallback {
+
+	/**
+	 * Attributes allowed on noscript fallback elements.
+	 *
+	 * This is used to prevent duplicated validation errors.
+	 *
+	 * @since 1.1
+	 * @var array
+	 */
+	private $noscript_fallback_allowed_attributes = array();
+
+	/**
+	 * Extends the default arguments array with additional defaults.
+	 *
+	 * @since 1.1
+	 *
+	 * @param array $default_args Original defaults.
+	 * @return array Extended defaults.
+	 */
+	protected function merge_default_args( array $default_args ) {
+		$default_args['add_noscript_fallback'] = true;
+		return $default_args;
+	}
+
+	/**
+	 * Initializes the internal allowed attributes array.
+	 *
+	 * @since 1.1
+	 *
+	 * @param string $tag Tag name to get allowed attributes for.
+	 */
+	protected function initialize_allowed_attributes( $tag ) {
```

Comment:
This is very specific to the noscript fallbacks, as it assumes there is it one tag spec for the given tag, which is true for the noscript fallback tag specs. In reality AMP has multiple tag specs for each tag name. So maybe this should also be renamed to be specific to the noscript fallback case.